onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,149 @@
+      dense_shape_(),
+      allocator_(),
+      location_(),
+      ml_data_type_(nullptr),
+      rep_() {
+}
+
+SparseTensor& SparseTensor::operator=(SparseTensor&& o) noexcept {
+  format_flags_ = o.format_flags_;
+  dense_shape_ = std::move(o.dense_shape_);
+  allocator_ = std::move(o.allocator_);
+  location_ = std::move(o.location_);
+  ml_data_type_ = std::move(o.ml_data_type_);
+  rep_ = std::move(o.rep_);
+  return *this;
+}
+
+SparseTensor::~SparseTensor() = default;
+
+SparseRep::SparseRep(MLDataType data_type, const TensorShape& values_shape,
+                     int64_t total_buffer_size, const AllocatorPtr& allocator)
+    : allocator_(allocator), 
+      p_data_(nullptr),
+      values_([&]() mutable -> Tensor {
+        // In the edge case when tensor is absolutely sparse, we do not have neither values, nor indices
+        if (total_buffer_size > 0) {
+          ORT_ENFORCE(values_shape.Size() * static_cast<int64_t>(data_type->Size()) < total_buffer_size,
+                      "Values size must be less than total buffer size");
+          auto data_ptr = IAllocator::MakeUniquePtr<void>(allocator_, static_cast<size_t>(total_buffer_size));
+          ORT_ENFORCE(data_ptr != nullptr, "SparseTensor Allocation failed for size: ", total_buffer_size);
+          // We own the buffer, so we must properly construct strings. Neither of the Tensors
+          // we construct on top of the buffer own it. We are constructing empty strings, hopefully
+          // nothrow and no buffer allocation
+          if (utils::IsDataTypeString(data_type)) {
+            const int64_t shape_size = values_shape.Size();
+            auto* ptr = static_cast<std::string*>(data_ptr.get());
+            for (int64_t i = 0; i < shape_size; ++i) {
+              new (ptr + i) std::string();
+            }
+          }
+          p_data_ = data_ptr.release();
+        }
+        return Tensor(data_type, values_shape, p_data_, allocator_->Info());
+      }()) {
+}
+
+SparseRep::SparseRep(MLDataType data_type, const TensorShape& values_shape,
+                     void* values, const OrtMemoryInfo& location)
+    : allocator_(),
+      p_data_(nullptr),
+      values_(data_type, values_shape, values, location) {
+}
+
+void SparseRep::ReleaseBuffer() {
+  if (allocator_ && p_data_ != nullptr) {
+    // if current tensor is responsible for deleting the buffer
+    // and it is a string tensor, need to explicitly call string(s)
+    // __dtor(s).
+    if (values_.IsDataTypeString()) {
+      using string = std::string;
+      auto* ptr = static_cast<std::string*>(p_data_);
+      const int64_t len = values_.Shape().Size();
+      for (int64_t i = 0; i < len; i++)
+        ptr[i].~string();
+    }
+    allocator_->Free(p_data_);
+  }
+}
+
+SparseRep::~SparseRep() {
+  ReleaseBuffer();
+}
+
+Status SparseTensor::Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const {
```

Comment:
>parseTensor::Cop [](http://example.com/codeflow?start=8&length=16)

Now that the data is in the single buffer, we can simply copy it for all primitive types and provide specialization for strings alone.